packages/docusaurus-theme-common/src/utils/codeBlockUtils.ts

```diff
@@ -99,50 +116,88 @@ export function parseLanguage(className: string): string | undefined {
  * Parses the code content, strips away any magic comments, and returns the
  * clean content and the highlighted lines marked by the comments or metastring.
  *
- * If the metastring contains highlight range, the `content` will be returned
- * as-is without any parsing.
+ * If the metastring contains a range, the `content` will be returned as-is
+ * without any parsing. The returned `lineClassNames` will be a map from that
+ * number range to the first magic comment config entry (which _should_ be for
+ * line highlight directives.)
  *
  * @param content The raw code with magic comments. Trailing newline will be
  * trimmed upfront.
- * @param metastring The full metastring, as received from MDX. Highlight range
- * declared here starts at 1.
- * @param language Language of the code block, used to determine which kinds of
- * magic comment styles to enable.
  */
 export function parseLines(
   content: string,
-  metastring?: string,
-  language?: string,
+  options: {
+    /**
+     * The full metastring, as received from MDX. Line ranges declared here
+     * start at 1.
+     */
+    metastring: string | undefined;
+    /**
+     * Language of the code block, used to determine which kinds of magic
+     * comment styles to enable.
+     */
+    language: string | undefined;
+    /**
+     * Magic comment types that we should try to parse. Each entry would
+     * correspond to one class name to apply to each line.
+     */
+    magicComments: MagicCommentConfig[];
+  },
 ): {
   /**
-   * The highlighted lines, 0-indexed. e.g. `[0, 1, 4]` means the 1st, 2nd, and
-   * 5th lines are highlighted.
+   * The highlighted lines, 0-indexed. e.g. `{0: ["highlight", "collapsible"]}`
+   * means the 1st line should have `highlight` and `collapsible as class names.
    */
-  highlightLines: number[];
+  lineClassNames: {[lineIndex: number]: string[]};
```

Comment:
Note: maybe we could make the system even more flexible so that a magic comment does not only add a className but could support as well arbitrary line props? 

This way it's possible to use it, swizzle `<Line>` and enhance it with custom JSX, without having to do `className.includes(myMagicCommentClass) && <div>MyDiv</div>`


Then the return could become: 

```ts
type LineConfig = {magicComments: MagicCommentConfig[]} // may be extended later?
 
lineConfigs: LineConfig[]
```

Or maybe just:

```ts 
lineProps: unknown[]
```

I think it's valuable to make the system extensible enough to support later more advanced cases, like:

```js
{
  line: 'pill-next-line',
  props: {pill: true, className: "code-pill"}
  parseCommentProps(comment) => ({
    pillNumber: parsePillNumber(comment), 
    pillColor: parsePillColor(comment)
  }), 
},
```

```js
// pill-next-line pillNumber=3 pillColor=red
const someLine = 42
```

Does it make sense?

Reply:
These things are ultimately passed through the theme config, so they have to be serializable. "class name" may not literally mean `className`—because it's an array of strings, if the user has swizzled code block, they can be free to do whatever they want with the class names.

`pill-next-line pillNumber=3 pillColor=red` seems a bit hard to parse, especially to do it robustly. I wouldn't do that for now until someone asks for it (I don't see an actual use-case: this can't be used for tooltips, for example)

Reply:
> These things are ultimately passed through the theme config, so they have to be serializable.

Oh yeah forgot about that 😅 

---

I've seen some docs where they render additional items like number badges in front of code lines (not necessarily ordered, 2 lines might even have the same number value). I thought I've seen this in the beta React doc, but can't find it anymore 🤷‍♂️ 

> pill-next-line pillNumber=3 pillColor=red seems a bit hard to parse, especially to do it robustly. I wouldn't do that for now until someone asks for it (I don't see an actual use-case: this can't be used for tooltips, for example)

I guess we could use a syntax akin to querystring eventually: `// pill-next-line pillNumber=3&pillColor=red`

But we can do that later